taca/element/Element_Runs.py

```diff
@@ -0,0 +1,1270 @@
+                                "PercentMismatch": percentage_mismatch,
+                                "PercentQ30": percentage_q30,
+                                "PercentQ40": percentage_q40,
+                                "QualityScoreMean": quality_score_mean,
+                            }
+                        )
+            else:
+                continue
+        return project_runstats
+
+    # Aggregate stats in IndexAssignment.csv
+    def aggregate_stats_assigned(self, demux_runmanifest):
+        aggregated_assigned_indexes = []
+        sub_demux_list = sorted(
+            list(set(sample["sub_demux_count"] for sample in demux_runmanifest))
+        )
+        lanes = sorted(list(set(sample["Lane"] for sample in demux_runmanifest)))
+        for sub_demux in sub_demux_list:
+            # Read in each Project_RunStats.json to fetch PercentMismatch, PercentQ30, PercentQ40 and QualityScoreMean
+            # Note that Element promised that they would include these stats into IndexAssignment.csv
+            # But for now we have to do this by ourselves in this hard way
+            project_runstats = self.get_project_runstats(sub_demux, demux_runmanifest)
+            # Read in IndexAssignment.csv
+            assigned_csv = os.path.join(
+                self.run_dir, f"Demultiplexing_{sub_demux}", "IndexAssignment.csv"
+            )
+            if os.path.exists(assigned_csv):
+                with open(assigned_csv) as assigned_file:
+                    reader = csv.DictReader(assigned_file)
+                    index_assignment = [row for row in reader]
+                for sample in index_assignment:
+                    if sample["Lane"] in lanes:
+                        project_runstats_sample = [
+                            d
+                            for d in project_runstats
+                            if d["SampleName"] == sample["SampleName"]
+                            and d["Lane"] == sample["Lane"]
+                            and d["ExpectedSequence"] == sample["I1"] + sample["I2"]
+                        ]
+                        sample["sub_demux_count"] = sub_demux
+                        sample["PercentMismatch"] = project_runstats_sample[0][
+                            "PercentMismatch"
+                        ]
+                        sample["PercentQ30"] = project_runstats_sample[0]["PercentQ30"]
+                        sample["PercentQ40"] = project_runstats_sample[0]["PercentQ40"]
+                        sample["QualityScoreMean"] = project_runstats_sample[0][
+                            "QualityScoreMean"
+                        ]
+                        aggregated_assigned_indexes.append(sample)
+            else:
+                logger.warning(
+                    f"No {os.path.basename(assigned_csv)} file found for sub-demultiplexing {sub_demux}."
+                )
+        # Remove redundant rows for PhiX
```

Comment:
Mostly out of curiosity: Where does the reduncancy come from? Is it from the `<project>_runstats.json`? @chuan-wang

Reply:
I was thinking about the cases that the same PhiX samples were included in multiple sub run manifests, with their indexes trimmed in different lengths.

Reply:
The cases that the same PhiX samples are included in multiple sub run manifests with different library types which requires different demux settings.